### PR TITLE
deprecate some contract api and support block prune

### PR DESCRIPTION
**Pull request by laizy**

### CODE REVIEW

File: core/store/ledgerstore/block_store.go [hunk 1]
```diff
@@ -473,45 +473,78 @@ func (this *BlockStore) getTransactionKey(txHash common.Uint256) []byte {
...
-func (this *BlockStore) getHeaderIndexListKey(startHeight uint32) []byte {
+func genHeaderIndexListKey(startHeight uint32) []byte {
 	sink := common.NewZeroCopySink(nil)
 	sink.WriteByte(byte(scom.IX_HEADER_HASH_LIST))
 	sink.WriteUint32(startHeight)
 	return sink.Bytes()
 }
 
-func (this *BlockStore) getStartHeightByHeaderIndexKey(key []byte) (uint32, error) {
+func getStartHeightByHeaderIndexKey(key []byte) (uint32, error) {
```
Comment (Honglei-Cong) [inline core/store/ledgerstore/block_store.go]: `getStartHeightByHeaderIndexKey` -> `getStartHeightFromHeaderIndexKey` ?

`By Key` seems reading from DB

File: core/store/ledgerstore/block_store.go [hunk 2]
```diff
@@ -473,45 +473,78 @@ func (this *BlockStore) getTransactionKey(txHash common.Uint256) []byte {
...
+
+func (this *BlockStore) SaveBlockPrunedHeight(height uint32) {
+	key := genBlockPruneHeightKey()
+	sink := common.NewZeroCopySink(nil)
+	sink.WriteUint32(height)
+
+	this.store.BatchPut(key, sink.Bytes())
+}
+
+func (this *BlockStore) PruneBlock(height uint32) {
```
Comment (Honglei-Cong) [inline core/store/ledgerstore/block_store.go]: duplicated with `SaveBlockPrunedHeight`?

File: core/store/ledgerstore/ledger_store.go [hunk 3]
```diff
@@ -1274,3 +1303,34 @@ func (this *LedgerStoreImp) Close() error {
...
+	if currHeader.Height <= config.GetContractApiDeprecateHeight() {
+		return 0
+	}
+	info, err := vconfig.VbftBlock(currHeader)
+	if err != nil {
+		return 0
+	}
+	lastReferHeight := info.LastConfigBlockNum
+	if info.NewChainConfig != nil {
+		lastReferHeight = currHeader.Height
```
Comment (Honglei-Cong) [inline core/store/ledgerstore/ledger_store.go:1331]: currHeader.Height - 1?

block which contains current chainconfig should be not pruned?

File: core/store/ledgerstore/ledger_store.go [hunk 4]
```diff
@@ -92,6 +92,8 @@ type LedgerStoreImp struct {
 	vbftPeerInfoblock    map[string]uint32 //pubInfo save pubkey,peerindex
 	lock                 sync.RWMutex
 	stateHashCheckHeight uint32
+
+	pruneBlock uint32 // block could be pruned if blockHeight + pruneBlock < currHeight , disable prune if equals 0
```
Comment (Honglei-Cong) [inline core/store/ledgerstore/ledger_store.go]: Is `pruneBlock` number of block which will not be pruned?

`pruneBlock` -> `preserveBlockHistoryLength`?

File: core/store/ledgerstore/ledger_store.go (hunk 3 above)
Comment (laizy) [reply core/store/ledgerstore/ledger_store.go:1331]: function returns `lastReferHeight -1`.

File: core/store/ledgerstore/block_store.go (hunk 1 above)
Comment (laizy) [reply core/store/ledgerstore/block_store.go]: it confused me too. so changed it as a free function. I will rename `get` to `gen` to make it more clear.

File: core/store/ledgerstore/ledger_store.go [hunk 5]
```diff
@@ -852,6 +854,33 @@ func (this *LedgerStoreImp) releaseSavingBlockLock() {
...
+		return false
+	}
+	if pruned+1 >= height {
+		return false
+	}
+
+	pruneHeight := pruned + 1
+	for ; pruneHeight-pruned < pruneBatchSize && pruneHeight < height; pruneHeight++ {
+		hash := this.GetBlockHash(pruneHeight)
+		this.blockStore.PruneBlock(hash)
```
Comment (Honglei-Cong) [inline core/store/ledgerstore/ledger_store.go]: does event-store need prune?

Comment (laizy) [reply core/store/ledgerstore/ledger_store.go]: done

**Comment by laizy** [comment]
@Honglei-Cong ping?

**Review by Honglei-Cong — APPROVED** [review]

**Pull request merged**